### PR TITLE
NLLB Tokenizer support.

### PR DESCRIPTION
**Pull request by npatsakula**
- [x] Vocabulary.
- [x] Tokenizer.
- [x] Tests.

### CODE REVIEW

**Comment by npatsakula** [comment]
Hello @guillaume-be! I made this separated PR on top of #78 and #80.

I'm already fixed issues from #76.

File: main/tests/test_nllb_cased.rs [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+        "eng_Latn This is a sample sentence to be tokénized",
+        "eng_Latn Wondering how this will get tokenized 🤔 ?",
+        "fra_Latn İs th!s 𩸽 Ϻ Šœ Ugljšić dấu nặng",
+        "hin_Deva   İs th!s    𩸽 Ϻ Šœ   Ugljšić  dấu nặng     ",
+        "lit_Latn � İs th!s �� 𩸽 Ϻ Šœ   Ugljšić  dấu nặng     ",
+    ];
+
+    let expected_results = [
+        TokenizedInput {
+            token_ids: vec![256127, 622, 2],
```
Comment (guillaume-be) [inline main/tests/test_nllb_cased.rs:42]: It looks like the EOS token (id 265047) is consistently missing when compared to the output from the Python tokenizer. The behaviour of the tokenizer should be identical to the Python version, the integration tests help capture such discrepancies

File: main/src/vocab/base_vocab.rs [hunk 2]
```diff
@@ -154,14 +154,15 @@ pub(crate) fn register_as_special_value(
...
-#[derive(Debug, Clone, Deserialize)]
+#[derive(Debug, Default, Clone, Deserialize)]
 pub struct SpecialTokenMap {
     pub unk_token: String,
     pub pad_token: Option<String>,
     pub bos_token: Option<String>,
     pub sep_token: Option<String>,
     pub cls_token: Option<String>,
     pub eos_token: Option<String>,
+    #[serde(skip_deserializing)]
```
Comment (guillaume-be) [inline main/src/vocab/base_vocab.rs:165]: This should probably be removed as `SpecialTokenMap` is used by all vocabs

File: main/src/vocab/nllb_vocab.rs [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+}
+
+impl NLLBVocab {
+    /// The beginning of sequence token that was used during pretraining.
+    /// Can be used a sequence classifier token.
+    pub fn get_bos_value(&self) -> &str {
+        self.special_token_storage
+            .bos_token
+            .as_deref()
+            .unwrap_or("<s>")
```
Comment (guillaume-be) [inline main/src/vocab/nllb_vocab.rs:77]: can you please align with https://github.com/guillaume-be/rust-tokenizers/blob/4d8aa3eb964fd33cb0dbc6bfd86b9edb5d264faf/main/src/vocab/m2m100_vocab.rs#L75 and declare the default constant separately - this would allow a clearer overview. Please also note it seems you are missing defaults for `cls`, `unk` and `mask`

File: main/src/vocab/nllb_vocab.rs [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+        self.special_token_storage
+            .eos_token
+            .as_deref()
+            .unwrap_or("</s>")
+    }
+
+    /// Returns the SEP token for M2M100 (`</s>`)
+    pub fn get_sep_value(&self) -> &str {
+        self.special_token_storage
+            .eos_token
```
Comment (guillaume-be) [inline main/src/vocab/nllb_vocab.rs:91]: should be `sep_token`

File: main/src/vocab/nllb_vocab.rs [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+
+    fn special_indices(&self) -> &HashMap<i64, String> {
+        &self.special_indices
+    }
+
+    fn from_file<V: AsRef<Path>>(vocab: V) -> Result<Self, TokenizerError> {
+        let tokenizer = Tokenizer::deserialize(vocab)?;
+        let special_config = SpecialTokenMap {
+            unk_token: "<unk>".into(),
+            ..Default::default()
```
Comment (guillaume-be) [inline main/src/vocab/nllb_vocab.rs:130]: This is dangerous as it would change the behaviour of the vocab if the shared `SpecialTokenMap` default changes for some reason. Can you please initialize as in https://github.com/guillaume-be/rust-tokenizers/blob/4d8aa3eb964fd33cb0dbc6bfd86b9edb5d264faf/main/src/vocab/m2m100_vocab.rs#L124

File: main/src/vocab/nllb_vocab.rs [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+            self.get_unknown_value(),
+        )
+    }
+
+    fn from_file_with_special_token_mapping<P: AsRef<Path>, S: AsRef<Path>>(
+        path: P,
+        special_token_mapping_path: S,
+    ) -> Result<Self, TokenizerError>
+    where
+        Self: Sized,
```
Comment (guillaume-be) [inline main/src/vocab/nllb_vocab.rs:159]: is the `Sized` specification required here?

File: main/src/vocab/nllb_vocab.rs [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+        let special_config = read_special_token_mapping_file(special_token_mapping_path)?;
+        Self::from_values_and_special_token_map(tokenizer_config.model.vocab, special_config)
+    }
+
+    fn from_values_and_special_token_map(
+        values: HashMap<String, i64>,
+        special_token_map: SpecialTokenMap,
+    ) -> Result<Self, TokenizerError>
+    where
+        Self: Sized,
```
Comment (guillaume-be) [inline main/src/vocab/nllb_vocab.rs:171]: can be skipped

File: main/src/vocab/nllb_vocab.rs [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+    {
+        let mut result = Self {
```
Comment (guillaume-be) [inline main/src/vocab/nllb_vocab.rs:173]: Could this be aligned with the other vocabularies instantiation for consistency? I believe this could look almost identical to the M2M100 vocab creation

File: main/src/vocab/nllb_vocab.rs [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+        result.indices = indices;
+        result.special_indices = special_indices;
+        result.special_values = special_values;
+        result.language_codes_bytes = language_code_bytes;
+        Ok(result)
+    }
+}
+
+#[derive(Deserialize)]
+struct Model {
```
Comment (guillaume-be) [inline main/src/vocab/nllb_vocab.rs:221]: Looks like this should be removed

**Review by guillaume-be — COMMENTED** [review]
Hello @npatsakula ,

Thank you for this PR. I left a few comments directly in the code. In addition to these, could you please add the Python bindings? These are especially useful for the addition of integration tests and validate the the output of the Python (from `transformers`) and Rust versions is identical

File: main/src/vocab/nllb_vocab.rs (hunk 9 above)
Comment (npatsakula) [reply main/src/vocab/nllb_vocab.rs:221]: This code is required for `tokenizer.json` deserialization (it nested and contains vocabulary + extra meta information).

File: main/src/vocab/base_vocab.rs (hunk 2 above)
Comment (npatsakula) [reply main/src/vocab/base_vocab.rs:165]: You're right, but structure of `special_tokens_map.json` is different for [NLLB](https://huggingface.co/facebook/nllb-200-3.3B/blob/main/special_tokens_map.json#L209) and [M2M](https://huggingface.co/facebook/m2m100_1.2B/blob/main/special_tokens_map.json#L1). I can avoid this skip using untagged deserializer, what do you think?

File: main/tests/test_nllb_cased.rs (hunk 1 above)
Comment (npatsakula) [reply main/tests/test_nllb_cased.rs:42]: Hm, looks like you're right. But I don't know why it's behave like this and this code works fine for my `rust-bert` translator.

Do you have any ideas why it's skipping EOS?

File: main/src/vocab/base_vocab.rs (hunk 2 above)
Comment (guillaume-be) [reply main/src/vocab/base_vocab.rs:165]: Thank you for pointing that out. It is unfortunate that models on the huggingface hub do not follow a set of conventions for the special token mask. For now I believe the best way forward would be to keep the general `SpecialTokenMap` unchanged, and indeed rely on a specific deserialization method for NLLB. You could also create a `NLLBSpecialTokenMap` that could deserialize the NLLB format, and then implement `From<NLLBSpecialTokenMap>` for `SpecialTokenMap`?

File: main/src/vocab/nllb_vocab.rs [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+
+    /// End of sequence token.
+    pub fn get_eos_value(&self) -> &str {
+        self.special_token_storage
+            .eos_token
+            .as_deref()
+            .unwrap_or("</s>")
+    }
+
+    /// Returns the SEP token for M2M100 (`</s>`)
```
Comment (guillaume-be) [inline main/src/vocab/nllb_vocab.rs:88]: Looks like a copy/paste error

File: main/src/vocab/nllb_vocab.rs [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+    pub fn get_sep_value(&self) -> &str {
+        self.special_token_storage
+            .eos_token
+            .as_deref()
+            .unwrap_or("</s>")
+    }
+
+    /// Returns the PAD token for M2M100 (`<pad>`)
```
Comment (guillaume-be) [inline main/src/vocab/nllb_vocab.rs:96]: invalid docstring

File: main/src/vocab/nllb_vocab.rs (hunk 9 above)
Comment (guillaume-be) [reply main/src/vocab/nllb_vocab.rs:221]: Please see the comment above. I think creating a `NLLBSpecialTokenMap` and `From<NLLBSpecialTokenMap>` for `SpecialTokenMap` would help keeping the original `SpecialTokenMap` unchanged and remove the need for a separate deserialization `Model` struct. What do you think?

File: main/tests/test_nllb_cased.rs (hunk 1 above)
Comment (guillaume-be) [reply main/tests/test_nllb_cased.rs:42]: I just checked and it is not skipping EOS (EOS is actually the token id `2`). The NLLB tokenizer has an extended suffix, as seen in https://github.com/huggingface/transformers/blob/2e35bac4e73558d334ea5bbf96a1116f7c0d7fb3/src/transformers/models/nllb/tokenization_nllb.py#L203 defined by [`eos`, `cur_lang_code`]. The `cur_lang_code` deaults to `eng_Latn`, and does not seem to be updated in most cases. I suggest the following:
- Add a `src_lang` field to the `NLLBTokenizer`, defaulting to `eng_Latn` at creation
- Add a `set_src_lang` method taking `(&mut self, src_lang)` as parameters to update this source language code if needed (do not expect to be useful in the standard case)
- Update the `build_input_with_special_tokens` to add the token, and mask/offset data after the EOS token

File: main/src/vocab/nllb_vocab.rs (hunk 9 above)
Comment (guillaume-be) [reply main/src/vocab/nllb_vocab.rs:221]: Just double-checked the file and you are correct - apologies did not realize that NLLB had a different vocabulary format altogether.

**Comment by guillaume-be** [comment]
Hello @npatsakula . It seems you are experiencing issues with the CI. I am now pushing the changes to the main branch which should resolve the issue (see https://github.com/guillaume-be/rust-tokenizers/pull/83). Please feel free to rebase.

Let me know if you need any further help for closing this PR and aligning the behaviour between Python and Rust, the NLLB inclusion would be a good addition to the crate.

File: .github/workflows/build.yml [hunk 12]
```diff
@@ -91,7 +91,7 @@ jobs:
       - run: |
           cd python-bindings
           pip install -r requirements-dev.txt
-          pip install transformers==4.12.3
+          pip install transformers==4.22.2
```
Comment (guillaume-be) [inline .github/workflows/build.yml:94]: You will want 4.21 (4.22 the download utilities used in the test and would have to be updated in a separate PR)

Comment (npatsakula) [reply .github/workflows/build.yml]: Yes, you're right, thank you very much!

Now it's failing during some of the `prophetnet` tests execution.

Comment (guillaume-be) [reply .github/workflows/build.yml]: Yes you'll want 
```python
            output_baseline.append(self.base_tokenizer.encode_plus(example.text_a,
                                                                   add_special_tokens=True,
                                                                   return_overflowing_tokens=True,
                                                                   return_special_tokens_mask=True,
                                                                   return_token_type_ids=True,
                                                                   max_length=128))
```
at https://github.com/guillaume-be/rust-tokenizers/blob/2913e21686446c093e8b216d83f0d4e647fdb839/python-bindings/tests/test_tokenization_sst2.py#L557

File: main/tests/test_nllb_cased.rs (hunk 1 above)
Comment (npatsakula) [reply main/tests/test_nllb_cased.rs]: Fixed in https://github.com/npatsakula/rust-tokenizers/pull/5.

File: main/src/vocab/base_vocab.rs (hunk 2 above)
Comment (npatsakula) [reply main/src/vocab/base_vocab.rs]: Fixed in https://github.com/npatsakula/rust-tokenizers/pull/5.

File: main/src/vocab/nllb_vocab.rs (hunk 3 above)
Comment (npatsakula) [reply main/src/vocab/nllb_vocab.rs]: Fixed in https://github.com/npatsakula/rust-tokenizers/pull/5.

File: main/src/vocab/nllb_vocab.rs (hunk 4 above)
Comment (npatsakula) [reply main/src/vocab/nllb_vocab.rs]: Fixed in https://github.com/npatsakula/rust-tokenizers/pull/5.

File: main/src/vocab/nllb_vocab.rs (hunk 6 above)
Comment (npatsakula) [reply main/src/vocab/nllb_vocab.rs:225]: You're right, it's redundant. Fixed in https://github.com/npatsakula/rust-tokenizers/pull/5.

File: main/src/vocab/nllb_vocab.rs (hunk 5 above)
Comment (npatsakula) [reply main/src/vocab/nllb_vocab.rs]: Fixed in https://github.com/npatsakula/rust-tokenizers/pull/5.

File: main/src/vocab/nllb_vocab.rs (hunk 11 above)
Comment (npatsakula) [reply main/src/vocab/nllb_vocab.rs]: Fixed in https://github.com/npatsakula/rust-tokenizers/pull/5.

File: main/src/vocab/nllb_vocab.rs (hunk 9 above)
Comment (npatsakula) [reply main/src/vocab/nllb_vocab.rs:294]: Workaround was added in https://github.com/npatsakula/rust-tokenizers/pull/5.

File: main/src/vocab/nllb_vocab.rs (hunk 10 above)
Comment (npatsakula) [reply main/src/vocab/nllb_vocab.rs]: Indeed. Fixed in https://github.com/npatsakula/rust-tokenizers/pull/5.

File: main/src/vocab/nllb_vocab.rs (hunk 8 above)
Comment (npatsakula) [reply main/src/vocab/nllb_vocab.rs:250]: Fixed in https://github.com/npatsakula/rust-tokenizers/pull/5.

File: .github/workflows/build.yml (hunk 12 above)
Comment (npatsakula) [reply .github/workflows/build.yml]: Pipeline fixed, thanx.

**Comment by npatsakula** [comment]
Hello @guillaume-be,

I resolved all issues, signed all commits and fixed all CI checks. I think this PR ready for review/merge. Feel free to contact me if there is something that I missed.

File: main/src/tokenizer/nllb_tokenizer.rs [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+};
+
+pub struct NLLBTokenizer {
+    model: SentencePieceBpeModel,
+    vocab: NLLBVocab,
+    src_lang: String,
+}
+
+impl NLLBTokenizer {
+    pub fn from_files<V: AsRef<Path>, M: AsRef<Path>, S: AsRef<Path>>(
```
Comment (guillaume-be) [inline main/src/tokenizer/nllb_tokenizer.rs:34]: I'm sorry for another small items: all tokenizers have 2 methods:
- `from_file(s)`: creation without special token map (using the defaults)
- `from_file(s)_with_special_token_map`: creation with a special token map

Could the methods please be aligned with the existing tokenizers?

**Review by guillaume-be — APPROVED** [review]
Thank for the PR!

I will publish this version to crates.io and push the necessary updates to the `rust-bert` crate this afternoon - this should then enable you to continue with the language model implementation on the associated pull request

File: main/src/tokenizer/nllb_tokenizer.rs (hunk 13 above)
Comment (npatsakula) [reply main/src/tokenizer/nllb_tokenizer.rs]: Done.

**Comment by npatsakula** [comment]
> Thank for the PR!
> 
> I will publish this version to crates.io and push the necessary updates to the `rust-bert` crate this afternoon - this should then enable you to continue with the language model implementation on the associated pull request

Great! I just want to notice that I've made PR for `rust-bert` that works in my environment for a long time:
https://github.com/guillaume-be/rust-bert/pull/279
